always return a promise; return last successful cache on error

FuzzyRhombus opened the pull request:
A promise is returned when caching is enabled. This makes it so a promise is always returned regardless of caching for more consistent handling of results.

Also if there is an error during compilation and a current valid cache is available that will be returned instead of failing

m90 commented:
Hey, thanks for your contribution. I like the idea of having an option to return stale cache results in case the least recent compilation failed, but I guess this should go into an option that is turned off by default - otherwise it'd be a. a breaking change and b. could lead to quite some headaches during debugging as it is not something I would expect as default behavior. If we put it into an option it'd be a cool addition though. Does this make sense to you?

Also, could you please make sure CI passes? Right now the tests won't even run due to linting issues.

m90 commented on index.js:
```diff
@@ -17,8 +17,7 @@ function lessExpress(location, lessOptions, options){
 	var localOptions = _.extend({}, globalOptions, options);
-	var localCache = localOptions.cache === false
-		? null
+	var localCache = localOptions.cache === false ? null
```
I think we do not need that change as it does the exact same thing as before.

FuzzyRhombus replied:
Only reason for this change was that for some reason it was keeping the tests from running for me. I'll give it a shot again after fixing the linting issues

m90 commented on index.js:
```diff
@@ -50,12 +49,20 @@ function lessExpress(location, lessOptions, options){
+			else throw err;
+		})
+		.catch(next);
+		return result;
```
I don't fully understand why you return something in here. This is the actual middleware, so the value returned from it will end up in limbo anyways. See the docs on middleware: http://expressjs.com/en/guide/using-middleware.html

FuzzyRhombus replied:
The reason I am returning the promise at the end is for the case of not being used directly as middleware, i.e. if there is going to be extra logic on the server before sending the result back. As you mentioned it is ignored as middleware so it should have no affect if it is never used.

In my case I have a general wrapper function used for multiple routes that wraps a given middleware using Promise.method and does some extra processing and error check before sending the result - this is why I am returning the promise, and also why the css is returned from the promise (line 57). It is not for general use case, but gives more flexibility. I hope this okay :)

m90 replied:
Ok, I get it. I am unsure if this should go in here though as it is not really clear why (if one does nit know about your particular use case). A middleware function explicitly exposes a `next` callback as its interface for asynchronicity. I'd feel it's quite odd to have it return a Promise doing the very same thing. In addition to that you could always wrap your "standard" express middleware yourself if you need it to return a Promise:

```js
function delayingMiddleware(req, res, next) {
	setTimeout(function() {
		let err;
		if (Math.random() > 0.9) err = new Error('Did not work');
		next(err);
	}, 500);
}

function wrapMiddleware(middlewareFn) {
	return function (req, res, next) {
		return new Promise(function(resolve, reject) {
			middlewareFn(req, res, function(err) {
				if (err) {
					reject(err);
				} else {
					resolve();
				}
				next(err);
			});
		});
	}
}

wrapMiddleware(delayingMiddleware);
```

Would handling it like this work for you?

FuzzyRhombus replied:
That is almost essentially what I am doing (though I am using `bluebird`s `Promise.method`). The issue with using `next` is you cannot pass along data - only errors. 
```javascript
function wrapMiddleware(middlewareFn) {
	return function (req, res, next) {
		return Promise.method(middlewareFn)(req, res, next)
                .then(function (returnedFromMiddleware) {
                        // do something else here
                        // errors are still handled using next
		});
	}
}
```
Now that I am reviewing it though, the response is only occurring when localCache is false/missing - whoops!

Mainly, my concern here is that there is no way to access the compiled css on the server - I can see this not being used often, so I understand why you are reluctant. Perhaps it would be more appropriate to omit the return statements and add the resulting css to `res.locals`:
```javascript
res.locals.lessCss = css;
res.set('Content-Type', 'text/css').send(css);
// return css; <- no more of this
```

This would eliminate returning the promise as well

Would that be acceptable? I think this way we are satisfying both of our concerns :)

m90 replied:
Ok, I see and I like the locals approach better than returning things as this will help others as well. Still I feel it's odd to attach sth to locals and then immediately end the response without any chance of ever accessing the locals. From what I understand you should need either the one (for inlining for example) or the other (sending a plain `text/css` resource)?

So, how about making this another option, sth along the lines of `passThru`:

```js
if (options.passThru){
  res.locals.lessCss = res.locals.lessCss || {};
  res.locals.lessCss[location] = css; 
  next();
} else {
  res.set('Content-Type', 'text/css').send(css);
}
```

Now we'd have an additional use case for the middleware:

```js
app.get('/style.css', lessExpress('./public/stylesheets/app.less'));
app.get('/landing,  lessExpress('./public/stylesheets/app.less', {}, {passThru: true}, function (req, res) {
  res.render('landing'); // => this would now have the css available in locals
});
```

Would that work for you?

FuzzyRhombus replied:
I think that will work just fine. I would still prefer to return the promise as it would allow cleaner syntax when wrapping the middleware (using `.method`) but I will leave it out for now; using `res.locals` would be a good approach either way. Give me some time and I will update the PR

m90 replied:
Ok, I guess even if the Promise based solution might make (a lot) more sense in your case, using a plain "express approach" in this case is the way to go as it will help interop and reusability.

Just ping me once you updated the PR and we can get this merged. Thanks.

m90 commented on index.js:
```diff
@@ -50,12 +49,20 @@ function lessExpress(location, lessOptions, options){
+		.catch(function (err) {
+			var lastBuild = localCache && localCache.get(location);
+			if (lastBuild) return lastBuild.then(function (css) {
+				res.set('Content-Type', 'text/css').send(css);
```
For the sake of being consistent I'd prefer if this wouldn't use chaining (like above), or if we use chaining in both cases.

FuzzyRhombus replied:
You're talking about the res chaining? Not a problem either way. I prefer chaining when it is still easily readable - if you are okay with changing each case to be chained. If not I will remove the chaining

m90 replied:
I have no preference, I'd just like it to be consistent, so feel free to pick whichever approach you prefer.

m90 commented on index.js:
```diff
@@ -50,12 +49,20 @@ function lessExpress(location, lessOptions, options){
 			}
 		}
 		result = render(location, localLessOptions).then(function(css){
+			if (localCache) localCache.set(location, result);
```
Why did you decide to do that before sending the response instead of doing that afterwards?

FuzzyRhombus replied:
No particular reason here. I can move it before the return if that is better. The reason for the return is explained above

FuzzyRhombus commented:
Thanks for the quick replies, I agree making this an option would be better. Coming up with a good name for the option, not so sure... `cacheErrorGuard` or something like that maybe? It would of course depend on `cache` not being false

m90 commented:
I think `stale` or `considerStale` would be a good option name.

FuzzyRhombus commented:
@m90 I've added the `stale` and `passThru` options like we talked about. All tests/CI should be passing. I realized a couple things in testing:
1. `return css;` (now line 65) is still needed for the LRU cache otherwise the cached promise has no result.
2. It seems whenever there is a change to a given location, that cache is disposed and no longer available for the stale case (not sure how it worked for me before... ¯\_(ツ)_/¯) . I've added a second cache object that simply adds a reference back to the promise when `dispose` is called

Take a look and let me know if you what you think

m90 commented on index.js:
```diff
@@ -17,13 +17,16 @@ function lessExpress(location, lessOptions, options){
 		paths: [_.initial(location.split('/')).join('/')]
 	}, globalLessOptions, lessOptions);
 	var localOptions = _.extend({}, globalOptions, options);
+	var localStale = localOptions.cache && localOptions.stale
```
Since this is used only a single time (and the var name isn't the most descriptive one) I think we can skip it can use `localOptions.cache && localOptions.stale` in Line 67?

m90 commented on test/fixtures/app.js:
```diff
@@ -14,6 +14,10 @@ app.get('/minified.css', lessExpress('./test/fixtures/minified.less', {plugins:
 app.get('/broken.css', lessExpress('./test/fixtures/broken.less'));
+app.get('/stale.css', lessExpress('./test/fixtures/stale.less', null, {cache: 1, stale: true}));
+app.get('/pass-thru.css', lessExpress('./test/fixtures/simple.less', null, {passThru: true}), function(req, res) {
+	res.status(202).send(res.locals.lessCss['./test/fixtures/simple.less']);
```
I like that approach to testing it :+1:

m90 commented on test/test.js:
```diff
@@ -93,5 +99,30 @@ describe('less-express', function(){
 				})
 				.end(done);
 		});
+		it('returns last cached css if less encounter an error', function(done){
```
Would it be possible to have the test clean up after itself, i.e. it could be run multiple times after itself?

FuzzyRhombus replied:
Sure, I can move the cleanup to the end of the test

m90 commented:
Looks pretty good now. Could you also add the new options to the README?

FuzzyRhombus commented:
@m90 Will do

FuzzyRhombus commented:
@m90 Ok, all updated

m90 commented:
Published as `less-express@1.1.0`. Thanks a whole lot for this :100: :100: :100:
